src/main/java/fi/hsl/transitdata/stopestimates/MetroEstimateStopEstimatesFactory.java

```diff
@@ -0,0 +1,118 @@
+        InternalMessages.StopEstimate.Builder builder = InternalMessages.StopEstimate.newBuilder();
+        InternalMessages.TripInfo.Builder tripBuilder = InternalMessages.TripInfo.newBuilder();
+
+        // TripInfo
+        tripBuilder.setTripId(metroEstimate.getDvjId());
+        tripBuilder.setOperatingDay(metroEstimate.getOperatingDay());
+        tripBuilder.setRouteId(metroEstimate.getRouteName());
+        tripBuilder.setDirectionId(Integer.parseInt(metroEstimate.getDirection()));
+        tripBuilder.setStartTime(metroEstimate.getStartTime());
+
+        // StopEstimate
+        if (metroEstimate.getJourneySectionprogress().equals(MetroAtsProtos.MetroProgress.CANCELLED)) {
+            return Optional.empty();
+        }
+        builder.setSchemaVersion(builder.getSchemaVersion());
+        builder.setTripInfo(tripBuilder.build());
+        builder.setStopId(metroEstimate.getStartStopNumber());
+        builder.setStopSequence(stopSequence);
+        // Status
+        switch (metroStopEstimate.getRowProgress()) {
+            case SCHEDULED:
+            case INPROGRESS:
+            case COMPLETED:
+                builder.setStatus(InternalMessages.StopEstimate.Status.SCHEDULED);
+                break;
+            case CANCELLED:
+                builder.setStatus(InternalMessages.StopEstimate.Status.SCHEDULED);
+                break;
+            default:
+                log.warn("Unrecognized status {}.", metroStopEstimate.getRowProgress());
+                break;
+        }
+        builder.setType(type);
+        // EstimatedTimeUtcMs & ScheduledTimeUtcMs
+        boolean isForecastMissing = false;
+        switch (type) {
+            case ARRIVAL:
+                if (!metroStopEstimate.getArrivalTimeForecast().isEmpty()) {
+                    builder.setEstimatedTimeUtcMs(ZonedDateTime.parse(metroStopEstimate.getArrivalTimeForecast()).toInstant().toEpochMilli());
+                    builder.setScheduledTimeUtcMs(ZonedDateTime.parse(metroStopEstimate.getArrivalTimePlanned()).toInstant().toEpochMilli());
+                } else {
+                    isForecastMissing = true;
+                }
+                break;
+            case DEPARTURE:
+                if (!metroStopEstimate.getDepartureTimeForecast().isEmpty()) {
+                    builder.setEstimatedTimeUtcMs(ZonedDateTime.parse(metroStopEstimate.getDepartureTimeForecast()).toInstant().toEpochMilli());
+                    builder.setScheduledTimeUtcMs(ZonedDateTime.parse(metroStopEstimate.getDepartureTimePlanned()).toInstant().toEpochMilli());
+                } else {
+                    isForecastMissing = true;
+                }
+                break;
+            default:
+                log.warn("Unrecognized type {}.", type);
+                break;
```

Comment:
Should we return `Optional.empty()` also if `type` is e.g. `null`?

Reply:
Added optional.empty to the other switch. Here it is not needed imo because if code gets to default: it will crash (because estimatedTimeUtcs is required field). Also it should crash to indicate bug because type is declared in this MetroEstimateStopEstimateFactory and it never go into default: